File: packages/gbnfjs/src/rules-builder/parse-char.ts
import { GrammarParseError, } from "../utils/errors.js";

export const parseChar = (src: string, pos: number): [number, number] => {
  if (src[pos] === '\\') {
    // if (src.slice(pos, pos + 1) === '\\') {
    switch (src[pos + 1]) {
      case 'x': return [parseInt(src.slice(pos + 2, pos + 4), 16), 4,];
      case 'u': return [parseInt(src.slice(pos + 2, pos + 6), 16), 6,];
      case 'U': return [parseInt(src.slice(pos + 2, pos + 10), 16), 10,];
      case 't': return ['\t'.charCodeAt(0), 2,];
      case 'r': return ['\r'.charCodeAt(0), 2,];
      case 'n': return ['\n'.charCodeAt(0), 2,];

      case '"':
      case '[':
      case ']':
        return [src.charCodeAt(pos + 1), 2,];
      case '\\':
        return [src.codePointAt(pos + 1), 2,];
      default:
        throw new GrammarParseError(src, pos, `Unknown escape at ${src[pos]}`);
    }
  }

  if (!src[pos]) {
    throw new GrammarParseError(src, pos, "Unexpected end of grammar input, failed to complete parse");
  }
  return [src.charCodeAt(pos), 1,];
};

File: packages/gbnfpy/gbnf/rules_builder/parse_char.py
from .errors import GrammarParseError


def parse_char(src: str, pos: int):
    if pos >= len(src):
        raise GrammarParseError(
            src,
            pos,
            "Unexpected end of grammar input, failed to complete parse",
        )

    if src[pos] == "\\":
        if src[pos + 1] == "x":
            return (int(src[pos + 2 : pos + 4], 16), 4)
        if src[pos + 1] == "u":
            return (int(src[pos + 2 : pos + 6], 16), 6)
        if src[pos + 1] == "U":
            return (int(src[pos + 2 : pos + 10], 16), 10)
        if src[pos + 1] in ["t", "r", "n"]:
            return (ord(eval(f"'\\{src[pos + 1]}'")), 2)
        if src[pos + 1] in ['"', "[", "]", "\\"]:
            return (ord(src[pos + 1]), 2)
        raise GrammarParseError(src, pos, f"Unknown escape at {src[pos]}")

    return (ord(src[pos]), 1)
